Eras in memory

afck opened the pull request:
Keep only two full eras in memory in the era supervisor, and one additional era for evidence validation only. See https://github.com/casper-network/casper-node/issues/1915 for details.

Closes https://github.com/casper-network/casper-node/issues/1915, https://github.com/casper-network/casper-node/issues/1588.

afck committed:
Remove EraSupervisorHandlingWrapper.

afck committed:
Fix era report/era end confusion.

afck committed:
Only keep three eras in memory.

xcthulhu commented on node/src/types/block.rs:
```diff
@@ -717,7 +717,7 @@ impl BlockHeader {
 
     /// Returns reward and slashing information if this is the era's last block.
-    pub fn era_end(&self) -> Option<&EraReport> {
+    pub fn era_report(&self) -> Option<&EraReport> {
```
👍

xcthulhu commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -181,47 +180,32 @@ where
+        let auction_delay = era_supervisor.protocol_config.auction_delay;
+        let last_activation_point = era_supervisor.protocol_config.last_activation_point;
+        let mut switch_blocks = Vec::new();
+        // We need to initialize current_era, current_era - 1 and current_era - 2. To initialize an
```
Consider changing to:

`// We need to initialize current_era, current_era - 1 and (frozen) current_era - 2. To initialize an`

goral09 replied:
Not sure if usage of word "frozen" makes it more clear. What does it even mean?

afck replied:
I'll add an explaining comment to the `active_eras` field:
So far I've used the term "evidence only" for this; it means that we deleted all the era's units and won't accept any new units, because we just keep the era around so we can verify evidence.

Not sure if "frozen" or "evidence only" is better?

goral09 replied:
Definitely prefer "evidence-only".

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/429a2b1af6acf011c0a141bc198e9bb52ecb75b8.

xcthulhu commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -181,47 +180,32 @@ where
+        // The create_new_era method initializes the era that the slice's last block is the key
+        // block for. We want to initialize the three latest eras, so we have to pass in the whole
+        // slice for the current era, and omit one or two elements for the other two. We never
+        // initialize the last_activation_point or an earlier era, however.
```
Consider a worked example with `Era(5)`, `Era(4)` and `Era(3)`

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/429a2b1af6acf011c0a141bc198e9bb52ecb75b8.

xcthulhu commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -278,124 +262,210 @@ where
+
+        // The beginning of the new era is marked by the key block.
+        #[allow(clippy::integer_arithmetic)] // Block height should never reach u64::MAX.
+        let start_height = key_block.height() + 1;
```
Add to `Block::verify` (or maybe `BlockHeader::verify`?) a check for `height < u64::MAX / 2`.

afck replied:
I'm still not sure about this: It's a very different kind of "invalid" than the hashes not matching.

E.g. if we asked for a block by hash, and its hashes don't match (current `verify` would fail), the fetcher should continue asking other peers for the block. But if the hashes do match and only the height is invalid, that means it should fail, and we should drop all blocks that referred to this one by hash.

goral09 commented on node/src/components/linear_chain_sync/operations.rs:
```diff
@@ -463,12 +462,9 @@ pub(crate) async fn run_fast_sync_task(
-    let historical_eras_needed = delay.saturating_mul(3).saturating_add(1);
+    // The era supervisor requires at least auction_delay + 3 eras back to be stored in the
+    // database.
+    let historical_eras_needed = chainspec.core_config.auction_delay.saturating_add(3);
```
Could we hide it behind `consensus::bonded_eras` (or similar) as well? I'm not a fan of hardcoding these assumptions all over the code. It's easy to miss and get out of sync.

afck replied:
I'm working on it. But I'll have to change the `EraSupervisor` to keep a `Chainspec` instead of a `ProtocolConfig`… that type should probably be removed anyway at some point.

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/429a2b1af6acf011c0a141bc198e9bb52ecb75b8.

afck committed:
Add more comments about eras in memory.

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -181,47 +180,32 @@ where
+        // and (excluding) current_era. However, we never use any block from before the last
+        // activation point.
+        let earliest_era =
+            last_activation_point.max(current_era.saturating_sub(auction_delay).saturating_sub(3));
```
Can we extract the `3` from the `saturating_sub` to a constant and give it a descriptive name?

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/c8b83472b3c9051e2e38ac016d940b33ecb2e37e.

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -181,47 +180,32 @@ where
+        for era_id in (earliest_era.value()..current_era.value()).map(EraId::from) {
+            let switch_block = storage
+                .read_switch_block_header_by_era_id(era_id)?
                 .ok_or_else(|| anyhow::Error::msg("No such switch block"))?;
```
Can you add an `era_id` to the error message please?

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/273073a47ece41dfdb75419b614c945b9fb11868.

afck committed:
Address review comments.

afck committed:
Remove consensus::ProtocolConfig.

afck committed:
Add era ID to error message.

afck committed:
Extract number of active eras into constant.

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -181,68 +190,47 @@ where
-                    .hash()
-            };
-            booking_blocks.insert(era_id, booking_block_hash);
+        // Collect the information needed to initialize all active eras.
+        //
+        // We need to initialize current_era, current_era - 1 and (evidence-only) current_era - 2.
+        // To initialize an era, all switch blocks between its booking block and its key block are
+        // required. The booking block for era N is in N - auction_delay - 1, and the key block in
+        // N - 1. So we need all switch blocks between:
+        // (including) current_era - 2 - auction_delay - 1 and (excluding) current_era.
+        // However, we never use any block from before the last activation point.
+        //
+        // Example: If auction_delay is 1, to initialize era N we need the switch blocks from era N
+        // and N - 1. If current_era is 10, we will initialize eras 10, 9 and 8. So we need the
+        // switch blocks from eras 9, 8, 7 and 6.
```
This is a really good explanation that will definitely make it easier for people to follow the code. Thanks 👍

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -298,124 +286,210 @@ where
+            warn!(era = era_id.value(), "era already exists");
+            return Effects::new();
+        }
+        if self.current_era > era_id.saturating_add(2) {
```
```suggestion
        if self.current_era > era_id.saturating_add(PAST_ACTIVE_ERAS) {
```

afck committed:
Add missed occurrence of PAST_ACTIVE_ERAS.

Co-authored-by: Mateusz Górski <goral09@users.noreply.github.com>

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -298,124 +286,210 @@ where
-        if let Some(obsolete_era_id) = oldest_evidence_era_id.checked_sub(1) {
+
+        // Remove the era that has become obsolete now: We keep only three in memory.
+        if let Some(obsolete_era_id) = self.current_era.checked_sub(3) {
```
Can we use `PAST_ACTIVE_ERAS + 1` here, instead of `3`?

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/f417ff94706385eda1d3d134cadcc7fec3f6dfd0.

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -723,32 +599,38 @@ where
             block_context,
         } = new_block_payload;
-        if !self.era_supervisor.is_bonded(era_id) {
+        if !self.active_eras.contains_key(&era_id) || era_id.successor() < self.current_era {
```
If we switched places of these statements, then the check could be slightly cheaper as comparing integers will be faster than checking set membership.

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/f417ff94706385eda1d3d134cadcc7fec3f6dfd0.

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -881,60 +701,68 @@ where
     fn has_evidence(&self, era_id: EraId, pub_key: PublicKey) -> bool {
-        self.era_supervisor
-            .iter_past(era_id, self.era_supervisor.bonded_eras())
+        self.iter_past(era_id, 1)
```
Can this `1` be extracted to a constant as well?

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/f417ff94706385eda1d3d134cadcc7fec3f6dfd0.

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -1116,53 +935,48 @@ where
-                        } else {
-                            continue;
-                        };
+                for e_id in self.iter_future(era_id, 1) {
```
Here we would use that new constant I left a comment about earlier.

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/f417ff94706385eda1d3d134cadcc7fec3f6dfd0.

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -1177,37 +991,96 @@ where
+/// Those are the booking block, i.e. the switch block in `era_id - auction_delay - 1`,
+/// the key block, i.e. the switch block in `era_id - 1`, and all switch blocks in between.
+async fn get_switch_blocks<REv>(
+    chainspec: Arc<Chainspec>,
```
Do we need a whole `Chainspec` or would that new type you had added suffice?

afck replied:
I _removed_ `ProtocolConfig`, and added a _trait_ `ChainspecConsensusExt`.

goral09 reviewed (approved):
Ńo blockers.

afck committed:
Replace hard-coded numbers with constants.

afck commented:
@goral09: Please double-check the latest commit; just want to make sure I didn't introduce any bugs with the constants. :sweat_smile:

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -67,8 +67,12 @@ pub use self::era::Era;
+const PAST_EVIDENCE_ERAS: u64 = 1;
 /// The number of active eras that are kept in memory in addition to the current one.
-const PAST_ACTIVE_ERAS: u64 = 2;
+const PAST_ACTIVE_ERAS: u64 = 2 * PAST_EVIDENCE_ERAS;
```
Is it really related?

afck replied:
I think so, yes:
We want to keep the past `PAST_EVIDENCE_ERAS` around as _full_ eras so that we can detect all the equivocations we would be allowed to cite.
Then, we have to keep a further `PAST_EVIDENCE_ERAS` in evidence-only mode because the earliest full era could cite evidence from them.

afck committed:
Call eras in memory "open", not "active".

The instantiated eras in memory include the evidence-only ones, that
don't have or accept any units anymore.

Commit:
Factored out `create_new_era` from `create_new_era_effects`

afck committed:
Fix build.

afck committed:
Add more sanity checks to create_new_era.

afck commented:
bors merge

afck commented:
(Failed due to `lsof` issue. retrying.)

afck commented:
bors merge

afck commented:
bors merge

afck commented:
Bors currently doesn't work with the `feat-fast-sync` branch. Merging manually.
